apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,174 @@
+ * strings representing individual errors. If there are no errors, returns an empty array.
+ */
+export function validatePCDCollection(pcds?: PCDCollection): string[] {
+  const validationErrors: string[] = [];
+
+  if (!pcds) {
+    validationErrors.push("pcd collection is absent");
+  }
+
+  if (pcds.size() === 0) {
+    validationErrors.push("pcd collection is empty");
+  }
+
+  const identityPCDFromCollection = pcds.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )[0] as SemaphoreIdentityPCD | undefined;
+
+  if (!identityPCDFromCollection) {
+    validationErrors.push("pcd collection does not contain an identity pcd");
+  }
+
+  return validationErrors;
+}
+
+export function validateUpload(user?: User, pcds?: PCDCollection): string[] {
+  const validationErrors: string[] = [];
+
+  if (!user) {
+    validationErrors.push(`upload must include a user`);
+  }
+
+  if (!pcds) {
+    validationErrors.push(`upload must include a pcd collection`);
+  }
+
+  const identityPCDFromCollection = pcds.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )[0] as SemaphoreIdentityPCD | undefined;
+  const commitmentFromPCDCollection =
+    identityPCDFromCollection?.claim?.identity?.commitment?.toString();
+
+  if (user?.commitment !== commitmentFromPCDCollection) {
+    validationErrors.push(
+      "user commitment does not equal to commitment of identity pcd in pcd collection"
+    );
+  }
+
+  return validationErrors;
+}
+
+/**
+ * Validates whether a user returned by the Zupass server API is consistent
+ * with the local application state representation. Returns errors as strings,
+ * and returns an empty array if the two are not inconsistent. Does not validate
+ * {@link state} in its entirety, only that the {@link user} and {@link state}
+ * are consistent.
+ */
+export function validateNewAccount(user: User, state: AppState): string[] {
+  const validationErrors: string[] = [];
+
+  if (!state.identity) {
+    validationErrors.push("app state missing identity field");
+  }
+
+  const stateIdentityCommitment = state.identity?.commitment?.toString();
+  const userIdentityCommitment = user?.commitment;
+
+  if (stateIdentityCommitment !== userIdentityCommitment) {
+    validationErrors.push(
+      `app state identity (${stateIdentityCommitment}) does not match newly created user's commitment (${userIdentityCommitment})`
+    );
+  }
+
+  return validationErrors;
+}
+
+/**
+ * Logs validation errors to the console, and uploads them to the server so that
+ * we have records and are able to identify common types of errors. Does not leak
+ * sensitive information, such as decrypted versions of e2ee storage.
+ */
+export async function logAndUploadValidationErrors(
+  errors: string[]
+): Promise<void> {
+  try {
+    console.log(`encountered state validation errors: `, errors);
+    await requestLogToServer(appConfig.zupassServer, "state-validation-error", {
+      errors
```

Comment:
Include UUID of the user, or some other way we could identify them.  Seems necessary to diagnose, as well as to understand whether we're seeing logs from many users, or for the same user repeatedly.

Also note I haven't reviewed the error messages in detail, though they should be checked for private user data, like semaphore keys.

Reply:
> Include UUID of the user, or some other way we could identify them. Seems necessary to diagnose, as well as to understand whether we're seeing logs from many users, or for the same user repeatedly.

I think that makes sense, though there are a few considerations:
- some of the error messages themselves include identifying information. i could see an argument for changing the mechanism by which I pass errors around to be an interface that has both some stable identifying information like a `uuid`, as well as a list of error messages. what do you think?
- some call-sites of some validation functions don't have access to the `appState`, and therefore don't have context on which user is calling them

> Also note I haven't reviewed the error messages in detail, though they should be checked for private user data, like semaphore keys.

If you mean check manually in this code review - that makes sense. If you mean we should come up with some automated way to check, I'm less sure about that though it is certainly possible. I was careful to not include sensitive information in the error messages.

Reply:
My goal with the uuid is to have something easy to filter by server-side, while not being too private.  Having it in all error reports in a specific JSON field (rather than inconsistently inside the message text) seems valuable.  I think an error reporting interface object may be a good way to package that information to pass down to layers which don't have access to it.  Or actually just stash it in a global variable may be good enough (so long as it's only for error reporting purposes).

It'll obviously have to be optional for logged-out scenarios, but those are limited.   Also it may even be valid for logged-out errors to be reported by the most recent known uuid if they happen soon after logging out.

I meant manually check the error messages for private data by review, yes.